awscrt/http_asyncio.py

```diff
@@ -0,0 +1,337 @@
+All network operations in `awscrt.http_asyncio` are asynchronous and use Python's asyncio framework.
+"""
+
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0.
+
+import asyncio
+from concurrent.futures import Future
+import awscrt.exceptions
+from typing import List, Tuple, Optional, Union, Callable, Any
+from awscrt.http import (
+    HttpClientConnection, HttpRequest, HttpClientStream, HttpProxyOptions,
+    Http2Setting, HttpVersion, Http2ClientStream
+)
+from awscrt.io import (
+    ClientBootstrap, SocketOptions, TlsConnectionOptions, InputStream
+)
+from collections import deque
+
+
+class HttpClientConnectionAsync(HttpClientConnection):
+    """
+    An async HTTP client connection.
+
+    Use `HttpClientConnectionAsync.new()` to establish a new connection.
+    """
+    __slots__ = ('_host_name', '_port')
+
+    @classmethod
+    async def new(cls,
+                  host_name: str,
+                  port: int,
+                  bootstrap: Optional[ClientBootstrap] = None,
+                  socket_options: Optional[SocketOptions] = None,
+                  tls_connection_options: Optional[TlsConnectionOptions] = None,
+                  proxy_options: Optional['HttpProxyOptions'] = None) -> "HttpClientConnectionAsync":
+        """
+        Asynchronously establish a new HttpClientConnectionAsync.
+
+        Args:
+            host_name (str): Connect to host.
+
+            port (int): Connect to port.
+
+            bootstrap (Optional [ClientBootstrap]): Client bootstrap to use when initiating socket connection.
+                If None is provided, the default singleton is used.
+
+            socket_options (Optional[SocketOptions]): Optional socket options.
+                If None is provided, then default options are used.
+
+            tls_connection_options (Optional[TlsConnectionOptions]): Optional TLS
+                connection options. If None is provided, then the connection will
+                be attempted over plain-text.
+
+            proxy_options (Optional[HttpProxyOptions]): Optional proxy options.
+                If None is provided then a proxy is not used.
+
+        Returns:
+            HttpClientConnectionAsync: A new HTTP client connection.
+        """
+        future = HttpClientConnection._generic_new(
+            host_name,
+            port,
+            bootstrap,
+            socket_options,
+            tls_connection_options,
+            proxy_options,
+            asyncio_connection=True)
+        return await asyncio.wrap_future(future)
+
+    async def close(self) -> None:
+        """Close the connection asynchronously.
+
+        Shutdown is asynchronous. This call has no effect if the connection is already
+        closing.
+
+        Returns:
+            None: When shutdown is complete.
+        """
+        close_future = super().close()
+        await asyncio.wrap_future(close_future)
```

Comment:
Generally you shouldn't change the type of an operation in a subclass. I would recommend either exposing a new method entirely (e.g. `close_async` or something) or you could just not subclass and instead wrap the sync client connection.